File: searchInsertPosition/searchInsertPosition.cpp
// Source : https://leetcode.com/problems/search-insert-position/
// Author : weekend27
// Date   : 2015-12-06

/**********************************************************************************

Given a sorted array and a target value, return the index if the target is found. If not, return the index where it would be if it were inserted in order.

You may assume no duplicates in the array.

Here are few examples.
[1,3,5,6], 5 → 2
[1,3,5,6], 2 → 1
[1,3,5,6], 7 → 4
[1,3,5,6], 0 → 0

**********************************************************************************/

// How to do it:
// binary search

class Solution {
public:
    int searchInsert(vector<int>& nums, int target) {
        int n = nums.size();
        if (n == 0)
            return 0;

        int l = 0, r = n - 1, m = 0;
        while (l <= r){
            m = l + (r - l) / 2;
            if (nums[m] == target)
                return m;
            else if (nums[m] > target)
                r = m - 1;
            else
                l = m + 1;
        }

    }
};

File: searchInsertPosition/searchInsertPosition.java
public class Solution {
    public int searchInsert(int[] nums, int target) {
        if (nums == null || nums.length == 0) {
            return 0;
        }

        if (target < nums[0]) {
            return 0;
        }
        else if (target > nums[nums.length - 1]) {
            return nums.length;
        }
        else {
            int left = 0;
            int right = nums.length - 1;
            while (left <= right) {
                int mid = left + (right - left) / 2;

                if (nums[mid] == target) {
                    return mid;
                } else if (target > nums[left] && target < nums[right] && (left == right - 1)) {
                    return right;
                } else if (nums[left] == target) {
                    return left;
                } else if (nums[right] == target) {
                    return right;
                } else if (nums[mid] > target) {
                    right = mid - 1;
                } else {
                    left = mid + 1;
                }
            }
        }

        return -1;
    }
}
